---
## Hausaufgabe 3 - Verbesserungsverfahren
---

### Achtung: Bitte überprüfen Sie vor Abgabe der Hausaufgabe, ob das Notebook richtig gespeichert wurde. 
Die Speicherung von Notebooks funktioniert die Menüleiste oder über die Shortcuts Strg + S (Win) und CMD + S (MAC).

---
## Bewertung
---

#### Erreichbare Punkte: 25
---
#### Erreichte Punkte:   24
---
---
Aufgabe 1: 2<br>
<br>
Aufgabe 2: 4<br>
<br>
Aufgabe 3: 18<br>

#### Aufgabe 1 - Theorie I: TwoEdgeExchange Move (2 Punkte)

Der 2-opt Algorithmus ist ein sehr bekanntes Verbesserungverfahren aus der Tourenplanung. Im Kern werden jeweils zwei **nicht-benachbarte** Kanten entfernt und die so entstandenen Teilsequenzen neu verbunden. Ein Beispiel ist in dem folgenden Bild dargestellt:

<img src="2-opt.png" alt="Drawing" style="width: 300px;"/>

Dieses Konzept kann auch auf das Scheduling angewandt werden. Dabei werden in einer Auftragsreihenfolge nicht nur die Aufträge als einzelne Elemente betrachtet, sondern auch deren Verbindungen. Wenn Auftrag $i$ direkt auf Auftrag $j$ folgt, so sind diese beiden Aufträge über eine Verbindungen, bzw. "Kante", $(i,j)$ verbunden. Somit lässt sich der sog. TwoEdgeExchange Move wie folgt definieren:

Ein TwoEdgeExchange Move $M(\text{2-EdgeExchange}, s, i, j)$ beschreibt den Austausch der Verbindungen zweier Auftragspaare $(i,i+1)$ und $(j,j+1)$. Dabei dürfen die Verbindungen **nicht-benachbart** sein und es werden **nur** die Verbindungen $(i,i+1)$ und $(j,j+1)$ entfernt und die Aufträge neu zu $(i,j)$ und $(i+1,j+1)$ zusammengefügt. 

Es ist folgende Auftragsreihenfolge gegeben `s=[0,1,2,3,4,5,6]`. Führen Sie (händisch) folgenden TwoEdgeExchange-Move durch $M(\text{2-EdgeExchange}, s, 1, 4)$.

**Hinweis:** Das erste Element steht an der Position 0.

$M(\text{2-EdgeExchange}, s, 1, 4) = [0, 1, 4, 3, 2, 5, 6]$

#### Aufgabe 2 - Theorie II: TwoEdgeExchange Move (5 Punkte)

**a)** In der Abbildung in Aufgabe 1 werden die Verbindungen zwischen den Auftragspaaren (0,1) und (2,3) ausgewählt. Nach den Autoren Kirlik und Oguz lässt sich jedoch der TwoEdgeExchange Move weiter einschränken, indem sich die Positionen $i$ und $j$ um mindestens 3 Position $\left(|j-i|\geq 3\right)$ unterscheiden müssen. Was könnte der Grund für diese weitere Einschränkung sein. 

**Hinweis:** Beachten Sie dabei die bereits umgesetzten Nachbarschaften.

**Literaturverweis** (in OPAL vorhanden): Kirlik, G., & Oguz, C. (2012). A variable neighborhood search for minimizing total weighted tardiness with sequence dependent setup times on a single machine. Computers & Operations Research, 39(7), 1506-1520.

## Antwort
Wenn i und j nur 1 unterscheiden, haben wir i+1 gleich j. Dann verändert die Sequenz nicht, es macht keinen Sinn. Bei heiristischen Verfahren möchten wir die Rechenzeit reduzieren.
Wenn i und j 2 unterscheiden, gibt es keine weitere Kanten zwischen i+1 und j. Die Auswirkung sieht gleich aus, wie wir einfach j und i+1 austauschen. Es macht weniger Sinn.
Deshalb beginnen wir mit abs(j-i) >= 3


<font color=red>
Korrektur</font> 

Wenn abs(j-i) = 2 gilt, gibt es die Kanten (i,i+1) und (j=i+2,j+1). Dieser Tausch erzeugt jedoch nur ein Vertauschen von i+1 und +2 (siehe Abbildung oben). Da dieser Tausch auch durch Swap oder Insertion erzeugt werden kann, kann dieser Tausch bei TwoEdgeExchange ausgeschlossen werden.

1 / 2 Punkte

**b)** Bestimmen Sie die Komplexität der TwoEdgeExchange-Move Nachbarschaft. Nehmen Sie an, dass eine Lösung durch eine Permutation der Länge $n$ repräsentiert ist und $|j-i|\geq 3$ gilt.

**Hinweise:** Geben Sie die Komplexität in Abhängigkeit von $n$ an.

**Tipp:**
<details>

Wie viele Kanten sind in einer Sequenz enthalten? Wie viele Kanten können getauscht werden, dass eine neue Lösung erzielt wird? Wieviele Dopplungen/Symmetrien gibt es?
</details>

## Antwort
$ |N(\text{2-EdgeExchange}, s)| = 1/2 * (n-4) * (n-3)$

Es gibt n Knoten, d.h. wir haben insgesamt n-1 Kanten.
Wir lassen zuerst i = 0, dann j beginnt erst mit j = 3. In diesem Fall haben wir noch (n-1)-3 = n-4 Möglichkeiten.
Wenn i = 1, haben wir noch (n-1)-1-3 = n-5 Möglichkeiten.
...
Am Ende haben wir nur 1 Möglichkeit. Wir brauchen wegen Symmetrie die Fälle von j < i nicht zu berucksichtigen.
Summieren: 1+2+3+...+(n-4)= 1/2 * (n-4) * (n-3)

<font color=red>
Korrektur</font> 

Sehr knapp, aber richtig hergeleitet -> 3/3 Punkte

#### Aufgabe 3 - Implementierung: TwoEdgeExchange-Move (18 Punkte)

Erweitern Sie den in der Lehrveranstaltung erstellten Solver um eine TwoEdgeExchange-Move Nachbarschaft. Erstellen Sie die Klassen `TwoEdgeExchangeMove` und `TwoEdgeExchangeNeighborhood`.



**Hinweise:** 

Implementieren Sie ihre Lösung in den bereitgestellten `.py` Dateien aus der Veranstaltung, nicht hier im Notebook. Bitte kopieren Sie der Übersichtlichkeit wegen trotzdem Ihre in der `.py`-Datei implementierten TwoEdgeExchange-Move Komponenten hier ins Notebook. Geben Sie alle zum fehlerfreien Ausführen notwendigen Dateien an.


Testen Sie die Nachbarschaft auf dem Datensatz `InputFlowshopSIST20.json`. Verwenden Sie die BestImprovement-Strategie und die FCFS-Startlösung. Mögliches erwartetes Ergebnis:


In [ ]:
# 2 Klassen in Datei Neighborhood
class TwoEdgeExchangeMove:
    def __init__(self, initialPermutation, indexA, indexB):
        self.Permutation = list(initialPermutation)
        self.IndexA = indexA
        self.IndexB = indexB

        list_left = initialPermutation[0: indexA+1]
        list_mid = reversed(initialPermutation[indexA+1: indexB+1])
        list_right = initialPermutation[indexB+1: ]

        self.Permutation = list_left+list(list_mid)+list_right

class TwoEdgeExchangeNeighborhood(BaseNeighborhood):
    def __init__(self, inputData, initialPermutation, evaluationLogic, solutionPool):
        super().__init__(inputData, initialPermutation, evaluationLogic, solutionPool)
        self.Type = 'TwoEdgeExchange'

    def DiscoverMoves(self):
        for i in range(0, len(self.Permutation)-4):
            for j in range(i+3, len(self.Permutation)-1):
                
                twoEdgeExchangeMove = TwoEdgeExchangeMove(self.Permutation, i, j)

                self.Moves.append(twoEdgeExchangeMove)

# ImprovementAlgorithm.py
def CreateNeighborhood(self, neighborhoodType, bestCurrentSolution):
    if neighborhoodType == 'Swap':
        return SwapNeighborhood(self.InputData, bestCurrentSolution.Permutation, self.EvaluationLogic, self.SolutionPool)
    elif neighborhoodType == 'Insertion':
        return InsertionNeighborhood(self.InputData, bestCurrentSolution.Permutation, self.EvaluationLogic, self.SolutionPool)
    elif neighborhoodType == 'TwoEdgeExchange':
        return TwoEdgeExchangeNeighborhood(self.InputData, bestCurrentSolution.Permutation, self.EvaluationLogic, self.SolutionPool)
    else:
        raise Exception(f"Neighborhood type {neighborhoodType} not defined.")

<font color=red>
Korrektur</font> 

Alles richtig umgesetzt!

In [1]:
from Solver import *


data = InputData("InputFlowshopSIST20.json")

localSearch = IterativeImprovement(data, 'BestImprovement', ['TwoEdgeExchange'])

solver = Solver(data, 1008)

solver.RunLocalSearch(
    constructiveSolutionMethod='FCFS',
    algorithm=localSearch)

Generating an initial solution according to FCFS.
Constructive solution found.
The permutation [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1482
The permutation [0, 1, 10, 9, 8, 7, 6, 5, 4, 3, 2, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1375
The permutation [0, 1, 10, 9, 8, 7, 6, 5, 4, 18, 17, 16, 15, 14, 13, 12, 11, 2, 3, 19] results in a Makespan of 1331
The permutation [0, 1, 10, 9, 8, 7, 6, 5, 4, 18, 17, 16, 15, 14, 13, 3, 2, 11, 12, 19] results in a Makespan of 1275
The permutation [0, 1, 10, 9, 8, 7, 6, 5, 4, 18, 15, 16, 17, 14, 13, 3, 2, 11, 12, 19] results in a Makespan of 1269
The permutation [0, 1, 10, 9, 8, 7, 15, 18, 4, 5, 6, 16, 17, 14, 13, 3, 2, 11, 12, 19] results in a Makespan of 1268
Reached local optimum of TwoEdgeExchange neighborhood. Stop local search.
Best found Solution.
The permutation [0, 1, 10, 9, 8, 7, 15, 18, 4, 5, 6, 16, 17, 14, 13, 3, 2, 11, 12, 19] results in a Makespan of 1268


          Generating an initial solution according to FCFS.
          Constructive solution found.
          The permutation [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1482
          The permutation [0, 1, 10, 9, 8, 7, 6, 5, 4, 3, 2, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1375
          The permutation [8, 9, 10, 1, 0, 7, 6, 5, 4, 3, 2, 11, 12, 13, 14, 15, 16, 17, 18, 19] results in a Makespan of 1360
          The permutation [8, 9, 10, 1, 0, 7, 6, 5, 4, 3, 2, 11, 12, 13, 14, 16, 15, 17, 18, 19] results in a Makespan of 1341
          The permutation [9, 8, 10, 1, 0, 7, 6, 5, 4, 3, 2, 11, 12, 13, 14, 16, 15, 17, 18, 19] results in a Makespan of 1340
          The permutation [9, 8, 10, 1, 0, 7, 6, 5, 4, 3, 2, 11, 12, 13, 14, 16, 17, 15, 18, 19] results in a Makespan of 1334
          The permutation [9, 8, 10, 1, 0, 7, 6, 5, 4, 12, 11, 2, 3, 13, 14, 16, 17, 15, 18, 19] results in a Makespan of 1312
          The permutation [9, 8, 10, 1, 0, 7, 6, 5, 4, 12, 17, 16, 14, 13, 3, 2, 11, 15, 18, 19] results in a Makespan of 1265
          The permutation [9, 1, 10, 8, 0, 7, 6, 5, 4, 12, 17, 16, 14, 13, 3, 2, 11, 15, 18, 19] results in a Makespan of 1256
          Reached local optimum of TwoEdgeExchange neighborhood. Stop local search.
          Best found Solution.
          The permutation [9, 1, 10, 8, 0, 7, 6, 5, 4, 12, 17, 16, 14, 13, 3, 2, 11, 15, 18, 19] results in a Makespan of 1256